first attempt at readout symmetrization

joshcombes opened the pull request:
first attempt at readout symm makes use of many of Kyle's functions like consolidate_symmetrization_outputs.

@kylegulshen It would be good for us to figure out what works for us and then get feedback from Matt and Sohaib.

There are lots of things to clean up but the main issue I see is integration with noise models.

The basic reason why it fails is as follows.
```python
prog = Progam(I(0))
prog += Noise...

qc.run_symmetrized_readout_new(prog, trials=160)

```
the `run_symmetrized_readout_new` will add symmetrization flips to the program after the noise which is useless. Conceptually the problem is the pyQuil abstraction is the noise is part of a program rather a property of the QVM.

To test the code I have added in a hack that adds noise in the `run_symmetrized_readout_new` method.

Commit:
first attempt at readout symm makes use of many of Kyle's functions like consolidate_symmetrization_outputs

kylegulshen commented on pyquil/api/_quantum_computer.py:
```diff
@@ -69,6 +72,289 @@ def _get_flipped_protoquil_program(program: Program) -> Program:
+
+
+def symmetrization(programs: List[Program], meas_qubits: List[List[int]], symm_type: str = 'thr') \
+        -> Tuple[List[Program], List[List[int]], List[Tuple[bool]], List[int]]:
```
```python
program: Program, meas_qubits: List[int], symm_type: int: int
```
symm_type = [-1, 0, 1, 2, 3] ...? 

```python
-> Tuple[List[Program], List[Tuple[bool]]
```
output only the list of ```symm_programs``` and the list of ```flip_arrays```


The point here is that we don't need to keep track of the origin program, since there is only one, and that the meas qubits won't change.

joshcombes replied:
- [x] to do

kylegulshen commented on pyquil/api/_quantum_computer.py:
```diff
@@ -69,6 +72,289 @@ def _get_flipped_protoquil_program(program: Program) -> Program:
+        elif symm_type == 'two':
+            flip_matrix = construct_strength_two_orthogonal_array(len(meas_qs))
+        elif symm_type == 'thr':
+            flip_matrix = construct_strength_three_orthogonal_array(len(meas_qs))
```
if the type is specified as int it may be easy to have a single construct_orthogonal_array(strenght: int, num_qubits: int)?

e.g. 
```python
if symm_type > 0:
    construct_orthogonal_array(symm_type, len(meas_qs))
```

Maybe it isn't simpler in the end though.

joshcombes replied:
Good idea

- [x] to do

kylegulshen commented on pyquil/api/_quantum_computer.py:
```diff
@@ -69,6 +72,289 @@ def _get_flipped_protoquil_program(program: Program) -> Program:
+    return symm_programs, symm_meas_qs, flip_arrays, prog_groups
+
+def consolidate_symmetrization_outputs(outputs: List[np.ndarray], flip_arrays: List[Tuple[bool]],
+                                       groups: List[int]) -> List[np.ndarray]:
```
You can remove '''groups''' since there will only ever be the one group (from the single ```program``` passed to ```symmetrization```).

```output``` below should just be a list.


This should return one np.ndarray

joshcombes replied:
- [x] to do

kylegulshen commented on pyquil/api/_quantum_computer.py:
```diff
@@ -69,6 +72,289 @@ def _get_flipped_protoquil_program(program: Program) -> Program:
+
+    return [np.vstack(output[group]) for group in sorted(list(set(groups)))]
+
+def _measure_bitstrings(qc, programs: List[Program], meas_qubits: List[List[int]],
```
```meas_qubits: List[int]```

remove ```assert```ion below and don't ```zip(programs, meas_qubits)```. Instead just make qubits -> meas_qubits

joshcombes replied:
- [x] to do

kylegulshen commented on pyquil/api/_quantum_computer.py:
```diff
@@ -189,6 +475,62 @@ def run_symmetrized_readout(self, program: Program, trials: int) -> np.ndarray:
+            raise ValueError("Symmetrization options are indicated by a length three string. See "
+                             "the docstrings for more information.")
+
+        if trials % 2 != 0:
```
won't this have to be more complicated? We need to either 
1) make trials the number of times each symm_program is run
2) round up the given trials so that they can be distributed evenly among different symm_programs

joshcombes replied:
This is a little more complicated. First I check for a power of two and then i check if there are enough trials.  It should be a little clearer now.

kylegulshen commented on pyquil/api/_quantum_computer.py:
```diff
@@ -189,6 +475,62 @@ def run_symmetrized_readout(self, program: Program, trials: int) -> np.ndarray:
+
+        if trials <= max_num_qubits:
+            trials = _next_power_of_2(2*max_num_qubits)
+            warnings.warn('Number of trials was too low it is now '+str(trials))
```
Related to comment above; I still think there would be something more needed here for different design strengths?

joshcombes replied:
yes you are right.

kylegulshen commented on pyquil/api/_quantum_computer.py:
```diff
@@ -189,6 +475,62 @@ def run_symmetrized_readout(self, program: Program, trials: int) -> np.ndarray:
+        if meas_qubits is None:
+            meas_qubits = [list(program.get_qubits())]
+
+        max_num_qubits= max( [len(qs) for qs in meas_qubits])
```
remove this? meas_qubits should just be List[int]

joshcombes replied:
- [x] to do

kylegulshen commented on pyquil/api/_quantum_computer.py:
```diff
@@ -189,6 +475,62 @@ def run_symmetrized_readout(self, program: Program, trials: int) -> np.ndarray:
+
+        conso_results = consolidate_symmetrization_outputs(results, flip_arrays, prog_groups)
+
+        return conso_results[0]
```
remove [0] after changing api above

joshcombes replied:
- [x] to do

Commit:
address kyle's suggestions on the symmetrization function and some other improvements

Commit:
construct_orthogonal_array and other fixes

Commit:
kyle's suggestions for consolidate_symm and _meas_bitstrings

Commit:
type annotations and docstrings + more cleanups

joshcombes commented:
Thanks for all the suggestions Kyle.

Based on Kyle's feedback I have cleaned up much of the code. I'll make a PR into pyquil to solicit fine grained feedback.

Commit:
start work on tests

Commit:
delete notebook and hacks that let me add noise

Commit:
some tests

Commit:
+scipy to requirements

Commit:
pylint

Commit:
small style changes

Commit:
docstrings

Commit:
address kyle's feedback

Commit:
small improvements

Commit:
code comment

Commit:
remove old code for run_symmetrized_readout and replace with new code

Commit:
remove un-needed tests and include more tests of the new code

Commit:
whitespace

Commit:
warning

Commit:
whitespace

Commit:
stronger testing

Commit:
test_construct_orthogonal_array

Commit:
test construct_strength_XX_orthogonal_array

Commit:
remove unused imports and make functions private

Commit:
remove helper functions from the top of the file

Commit:
add helper functions to the bottom of the file

Commit:
address Marcus' suggestion i.e. remove reference to "design"

Commit:
Merge remote-tracking branch 'upstream/master' into revamp-readout-sym

Commit:
Merge remote-tracking branch 'origin/master' into revamp-readout-sym

Commit:
more tests

Commit:
temporarily added the notebook back in at Sohaib request, skipped the tests that fail when i hardcode noise models in.

Commit:
whitespace

Commit:
remove scipy dependence

Commit:
Address Peter's suggestions

appleby committed:
Fix mathjax macros

Move the mathjax macros out of _static/mathjax_macros.js and into
Sphinx's mathjax_config variable.

Prior to this change, the macros were failing to load for me, which
caused them to be rendered literally ("\ket") in a garish red font.

When I looked in Chrome Developer tools console, it complained:

    Uncaught ReferenceError: MathJax is not defined
        at mathjax_macros.js:1

I think this is due to the fact that Sphinx loads MathJax
asynchronously from a CDN, then loads mathjax_macros.js immediately
after, so it's possible the macros get loaded before MathJax is
defined. Prior to this change, Sphinx was generating HTML like so:

    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <script type="text/javascript" src="_static/mathjax_macros.js"></script>

Additionally, I believe the path specified as the argument to
MathJax.Ajax.loadComplete at the bottom of mathjax_macros.js was
wrong. It specified a path of "[MathJax]/config/local/local.js", but I
think should have been something like "_static/mathjax_macros.js".

After this change, the macros config is injected inline on every page
by Sphinx, alongside the other MathJax config, like so:

    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"TeX": {"Macros": {"sket": ["\\left|\\left. #1 \\right\\rangle\\!\\right\\rangle", 1], "sbra": ["\\left\\langle\\!\\left\\langle #1 \\right.\\right|", 1], "sbraket": ["\\left\\langle\\!\\left\\langle #1 | #2 \\right\\rangle\\!\\right\\rangle", 2], "ket": ["\\left| #1 \\right\\rangle", 1], "bra": ["\\left\\langle #1 \\right|", 1], "braket": ["\\left\\langle #1 | #2 \\right\\rangle", 2], "vec": ["\\text{vec}\\left(#1\\right)", 1], "tr": ["\\text{Tr}\\left(#1\\right)", 1]}}, "tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>

This ensures that the macros get loaded along with the rest of the
config whenever the async MathJax load from the CDN completes.

Sphinx Doc References:

https://www.sphinx-doc.org/en/master/usage/extensions/math.html#confval-mathjax_config

https://docs.mathjax.org/en/latest/configuration.html#using-in-line-configuration-options

joshcombes committed:
Persistent initial state in reference density simulator (#920)

Commit:
Merge remote-tracking branch 'upstream/master' into revamp-readout-sym

# Conflicts:
#	pyquil/tests/test_quantum_computer.py

Commit:
remove exploratory code and notebook

notmgsk committed:
Use native Gate.dagger() in Program.dagger() (#887)

This removes a bunch of logic around creating inverse gates from
user-defined gates. If anyone thinks that stuff is useful, then I
would argue that it should be in some util function, not in
Program.dagger().

appleby committed:
Fix inlinemath rendering in intro.rst + other small doc typos (#927)

karalekas committed:
Add note about changelog to the PR template

karalekas committed:
Update changelog for 2.9.0

karalekas committed:
Bump version to 2.9.0

karalekas committed:
Relax quilc server requirement when using QVM / pyQVM (#934)

karalekas committed:
Bump version to 2.9.1

appleby committed:
Fix sphinx warnings (#929)

* Fix intersphinx warning
* Fix Wavefunction.amplitudes warning
* Fix inline literal start-string without end-string warning
* Fix field list ends without a blank line warning
* Fix unexpected section title warning
* Fix substitution_reference warnings
* Fix title level inconsistent warning
* Fix file not found warning
* Fix document isn't included in any toctree warning
* Fix undefined label warnings
* Fix could not lex literal_block as "python" warning
* Fix duplicate object description warning
* Tell sphinx-build to treat warnings as errors
* Add a clean target to the docs Makefile

joshcombes committed:
Merge pull request #8 from rigetti/master

update

Commit:
replace * with .

Commit:
Merge branch 'revamp-readout-sym' of https://github.com/joshcombes/pyquil into revamp-readout-sym

Commit:
replace { and } with ( and )

Commit:
docstring

Commit:
trivial commit to trigger build because of stochastically failing test

kylegulshen committed:
Support for dagger when converting gate to matrix (#931)

karalekas committed:
Add markdown version of the changelog

karalekas committed:
Add PyPI & conda version shields.io badges to README (#946)

karalekas committed:
Delete deprecated modules job_results.py and kraus.py

karalekas committed:
Clean up and refresh the README (#949)

msohaibalam committed:
Optional faster version of the operator estimation UTs (#928)

notmgsk committed:
Modernize example notebooks (#820)

notmgsk committed:
Toggle protoquil compilation from compiler client (#940)

starktech23 committed:
Update QCS and Slack invite links in index.rst (#965)

karalekas committed:
Bump rpcq dependency version to v2.7.2

notmgsk committed:
Add docs for program time estimator (#963)

karalekas committed:
Bump version to 2.10.0

maxKenngott committed:
Fix MemoryReference warnings in tests (#950)

kylegulshen committed:
Add merge_programs support for DefPermutationGate (#971)

karalekas committed:
Update the changelog to reflect v2.11 progress

karalekas committed:
Fix extra indent in v2.10 section

karalekas committed:
Update markdown changelog to match master

karalekas committed:
Fix some missing double quotes

karalekas committed:
Fix line length for PR template link

karalekas committed:
Remove RTD-only constructs from the rst changelog

karalekas committed:
Regenerate the md changelog without RTD-only constructs

karalekas committed:
Use level-3 headings in the md changelog

karalekas committed:
Put the @ symbol before github usernames in md changelog

karalekas committed:
Fix one missing level-3 header

karalekas committed:
Use pandoc to regenerate the rst changelog from md one

karalekas committed:
Update PR template with new changelog instructions

karalekas committed:
Update the changelog with an announcement about this PR

karalekas committed:
Add GH user handles to changelog for last couple releases

karalekas committed:
Add pandoc to CI for changelog rst->md conversion

karalekas committed:
Link the release headings to source code diffs

karalekas committed:
Merge branch 'master' into revamp-readout-sym
